internal/querynodev2/delegator/idf_oracle.go

```diff
@@ -0,0 +1,260 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package delegator
+
+import (
+	"fmt"
+	"sync"
+
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus-proto/go-api/v2/commonpb"
+	"github.com/milvus-io/milvus/internal/querynodev2/segments"
+	"github.com/milvus-io/milvus/internal/storage"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+)
+
+type IDFOracle interface {
+	// Activate(segmentID int64, state commonpb.SegmentState) error
+	// Deactivate(segmentID int64, state commonpb.SegmentState) error
+
+	SyncDistrbution(snapshot *snapshot)
+
+	UpdateGrowing(segmentID int64, stats map[int64]*storage.BM25Stats)
+
+	Register(segmentID int64, stats map[int64]*storage.BM25Stats, state commonpb.SegmentState)
+	Remove(segmentID int64, state commonpb.SegmentState)
+
+	BuildIDF(fieldID int64, tfs ...map[uint32]float32) ([][]byte, float64, error)
+}
+
+type bm25Stats struct {
+	stats         map[int64]*storage.BM25Stats
+	activate      bool
+	targetVersion int64
+}
+
+func (s *bm25Stats) Merge(stats map[int64]*storage.BM25Stats) {
+	for fieldID, newstats := range stats {
+		if stats, ok := s.stats[fieldID]; ok {
+			stats.Merge(newstats)
+		} else {
+			s.stats[fieldID] = newstats.Clone()
+		}
+	}
+}
+
+func (s *bm25Stats) Diff(stats map[int64]*storage.BM25Stats) {
+	for fieldID, newstats := range stats {
+		if stats, ok := s.stats[fieldID]; ok {
+			stats.Diff(newstats)
+		} else {
+			log.Panic("diff for empty bm25 stats")
+		}
+	}
+}
+
+func (s *bm25Stats) GetStats(fieldID int64) (*storage.BM25Stats, error) {
+	stats, ok := s.stats[fieldID]
+	if !ok {
+		return nil, fmt.Errorf("field not found in bm25 stats")
+	}
+	return stats, nil
+}
+
+func (s *bm25Stats) NumRow() int64 {
+	for _, stats := range s.stats {
+		return stats.NumRow()
+	}
+	return 0
+}
+
+func newBm25Stats() *bm25Stats {
+	return &bm25Stats{
+		stats: make(map[int64]*storage.BM25Stats),
+	}
+}
+
+type idfOracle struct {
+	sync.RWMutex
+
+	current *bm25Stats
+
+	growing map[int64]*bm25Stats
+	sealed  map[int64]*bm25Stats
+
+	targetVersion int64
+}
+
+func (o *idfOracle) Register(segmentID int64, stats map[int64]*storage.BM25Stats, state commonpb.SegmentState) {
+	o.Lock()
+	defer o.Unlock()
+
+	log.Info("test-- register idf", zap.Int64("segmentID", segmentID), zap.String("state", state.String()))
+	switch state {
+	case segments.SegmentTypeGrowing:
+		if _, ok := o.growing[segmentID]; ok {
+			return
+		}
+		o.growing[segmentID] = &bm25Stats{
+			stats:         stats,
+			activate:      true,
+			targetVersion: initialTargetVersion,
+		}
+		o.current.Merge(stats)
+	case segments.SegmentTypeSealed:
+		if _, ok := o.sealed[segmentID]; ok {
+			return
+		}
+		o.sealed[segmentID] = &bm25Stats{
+			stats:         stats,
+			activate:      false,
+			targetVersion: initialTargetVersion,
+		}
```

Comment:
should we call `o.current.Merge(stats)`?

Reply:
No, sealed segment should be deactivated till sync distribution.